### PR TITLE
cohttp: move new client and server modules into a generic module

### PR DESCRIPTION
**Pull request by mseri**
to avoid shadowing Client and Server modules when opening Cohttp.

This preserves backward compatibility for a very small price and improves library ergonomics since right now many project tend to open Cohttp to have the Headers and Response modules directly accessible.

### CODE REVIEW

**Comment by mseri** [comment]
Ping @samoht @mefyl @talex5.
It makes sense to have them as Cohttp.Client and Cohttp.Server, but it is true that I also have a number of scripts which `open Cohttp` for convenience. What do you think?

I have seen this failure in the opam-repository revdeps:
```
#=== ERROR while compiling prometheus-app.1.2 =================================#
# context              2.2.0~alpha3~dev | linux/x86_64 | ocaml-base-compiler.4.14.1 | file:///home/opam/opam-repository
# path                 ~/.opam/4.14/.opam-switch/build/prometheus-app.1.2
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p prometheus-app -j 255
# exit-code            1
# env-file             ~/.opam/log/prometheus-app-7-9681f8.env
# output-file          ~/.opam/log/prometheus-app-7-9681f8.out
### output ###
# (cd _build/default && /home/opam/.opam/4.14/bin/ocamlopt.opt -w -40 -g -I app/.prometheus_app.objs/byte -I app/.prometheus_app.objs/native -I /home/opam/.opam/4.14/lib/angstrom -I /home/opam/.opam/4.14/lib/asetmap -I /home/opam/.opam/4.14/lib/astring -I /home/opam/.opam/4.14/lib/base64 -I /home/opam/.opam/4.14/lib/bigstringaf -I /home/opam/.opam/4.14/lib/cohttp -I /home/opam/.opam/4.14/lib/cohttp-lwt -I /home/opam/.opam/4.14/lib/fmt -I /home/opam/.opam/4.14/lib/http -I /home/opam/.opam/4.14/lib/http/__private__/http_bytebuffer -I /home/opam/.opam/4.14/lib/logs -I /home/opam/.opam/4.14/lib/lwt -I /home/opam/.opam/4.14/lib/ppx_sexp_conv/runtime-lib -I /home/opam/.opam/4.14/lib/prometheus -I /home/opam/.opam/4.14/lib/re -I /home/opam/.opam/4.14/lib/seq -I /home/opam/.opam/4.14/lib/sexplib0 -I /home/opam/.opam/4.14/lib/stringext -I /home/opam/.opam/4.14/lib/uri -I /home/opam/.opam/4.14/lib/uri-sexp -intf-suffix .ml -no-alias-deps -o app/.prometheus_app.objs/native/prometheus_app.cmx -c -impl app/prometheus_app.ml)
# File "app/prometheus_app.ml", line 157, characters 6-27:
# 157 |       Server.respond_string ~status:`OK ~headers ~body ()
#             ^^^^^^^^^^^^^^^^^^^^^
# Error: Unbound value Server.respond_string
# (cd _build/default && /home/opam/.opam/4.14/bin/ocamlc.opt -w -40 -g -bin-annot -I app/.prometheus_app.objs/byte -I /home/opam/.opam/4.14/lib/angstrom -I /home/opam/.opam/4.14/lib/asetmap -I /home/opam/.opam/4.14/lib/astring -I /home/opam/.opam/4.14/lib/base64 -I /home/opam/.opam/4.14/lib/bigstringaf -I /home/opam/.opam/4.14/lib/cohttp -I /home/opam/.opam/4.14/lib/cohttp-lwt -I /home/opam/.opam/4.14/lib/fmt -I /home/opam/.opam/4.14/lib/http -I /home/opam/.opam/4.14/lib/logs -I /home/opam/.opam/4.14/lib/lwt -I /home/opam/.opam/4.14/lib/ppx_sexp_conv/runtime-lib -I /home/opam/.opam/4.14/lib/prometheus -I /home/opam/.opam/4.14/lib/re -I /home/opam/.opam/4.14/lib/seq -I /home/opam/.opam/4.14/lib/sexplib0 -I /home/opam/.opam/4.14/lib/stringext -I /home/opam/.opam/4.14/lib/uri -I /home/opam/.opam/4.14/lib/uri-sexp -intf-suffix .ml -no-alias-deps -o app/.prometheus_app.objs/byte/prometheus_app.cmo -c -impl app/prometheus_app.ml)
# File "app/prometheus_app.ml", line 157, characters 6-27:
# 157 |       Server.respond_string ~status:`OK ~headers ~body ()
#             ^^^^^^^^^^^^^^^^^^^^^
# Error: Unbound value Server.respond_string
```

**Review by talex5 — APPROVED** [review]
Fine with me. The cohttp README itself uses `open cohttp` in the examples, and presumably only works now because of the order of the opens.

**Comment by mseri** [comment]
We don't test the readme directly afaik, so it may as well be broken

**Comment by mefyl** [comment]
tldr: IMO fine in 5.0.0

AFAIK there is no consensus in the Ocaml community regarding backward compatibility and versioning and in my opinion we really need one. Some developers / company have their own approach, but most of the time I don't feel like there is any guarantee as to what package versions to pin to avoid breakage, and I tend to err on the paranoid side.

My personal take on this that the absolute worse thing is compilation and CIs breaking one morning because a new version of an external dependency was released. More so since this often does not happen immediately because build images are often cached or use lockfile, and will happen randomly when the image or lockfile is regenated. It follows that we need a clear guarantee of no API breakage depending on the versionning, and if we're going with semver, which is not perfect but is the most universally adopted thing, that means no API breakage except for major updates, so my opinion on that is that this is fine for cohttp 5 only. I think it's reasonable to expect some possible breakage if you jump major versions, and if you don't want that you can (should) `(cohttp (and (>= 4.0.0) (< 5.0.0)))`.

On another note there are ways to have API change on minor versions while providing backward compatibility modules, having the better of both worlds provided some additional work for the maintainer. We do this internally quite successfully, and I've seen other do. I don't know if anyone ever wrote down a ruleset, but I'd be happy to submit how we proceed for comments.

**Comment by mseri** [comment]
> We agree then, and alpha/betas are of course fine, the name imply you will encounter rough edges.

Absolutely! I suggested this since I think it is a usability improvement, I didn't mean to suggest that we had to keep backward compatibility at all costs
